assets/src/domain/eventEditor/services/apollo/mutations/useReorderEntities.ts

```diff
@@ -0,0 +1,127 @@
+import { useCallback, useEffect, useState } from 'react';
+import { useMutation } from '@apollo/react-hooks';
+import gql from 'graphql-tag';
+import { clone, reverse } from 'ramda';
+import { useDebouncedCallback } from 'use-debounce';
+
+import { EntityId } from '@appServices/apollo/types';
+import { getGuids } from '@appServices/predicates';
+import { Datetime, Ticket } from '../types';
+
+type Entity = Datetime | Ticket;
+
+interface ReorderEntitiesProps {
+	entityType: 'DATETIME' | 'TICKET';
+}
+
+interface CallbackArgs<E extends Entity> {
+	allEntities: Array<E>;
+	filteredEntities: Array<E>;
+	newIndex: number;
+	oldIndex: number;
+	updateEntityList: (updatedEntities: Array<E>) => void;
+}
+
+type SortCallback<E extends Entity> = (args: CallbackArgs<E>) => void;
+
+const REORDER_ENTITIES = gql`
+	mutation REORDER_ENTITIES($input: ReorderEspressoEntitiesInput!) {
+		reorderEspressoEntities(input: $input) {
+			ok
+		}
+	}
+`;
+
+interface ReorderEntities<E extends Entity> {
+	cancel: VoidFunction;
+	done: VoidFunction;
+	sortEntities: SortCallback<E>;
+}
+
+const useReorderEntities = <E extends Entity>({ entityType }: ReorderEntitiesProps): ReorderEntities<E> => {
+	const [allEntityGuids, setAllEntityGuids] = useState<Array<EntityId>>([]);
+	useEffect(() => {
+		if (allEntityGuids.length) {
+			done();
+		}
+	}, [allEntityGuids.join(':')]);
+
+	const [mutate, result] = useMutation(REORDER_ENTITIES);
+	useEffect(() => {
+		console.log(result);
+	}, [result]);
+
+	const [runMutation, cancelDebounce] = useDebouncedCallback((...args) => {
+		console.log('runMutation called');
+		mutate(...args);
+	}, 5000); // delay in MS
+
+	const done = useCallback(() => {
+		runMutation({
+			variables: {
+				input: {
+					clientMutationId: 'REORDER_ENTITIES',
+					entityIds: allEntityGuids,
+					entityType,
+				},
+			},
+		});
+	}, [allEntityGuids, entityType, runMutation]);
+
+	const cancel = useCallback(() => {
+		cancelDebounce();
+	}, [cancelDebounce]);
+
+	const sortEntities = useCallback<SortCallback<E>>(
+		({ allEntities: allEntitiesList, filteredEntities, newIndex, oldIndex, updateEntityList }) => {
+			if (newIndex === oldIndex || newIndex < 0 || oldIndex < 0) {
+				return;
+			}
+			// cancel existing debounce
+			cancel();
+
+			const entities = clone(filteredEntities);
+			const allEntities = clone(allEntitiesList);
+
+			// remove entity from existing location in filtered list
+			const [removed] = entities.splice(oldIndex, 1);
```

Comment:
it's good that you cloned this array before because `splice` is mutating it

Reply:
No, `Array.slice()` does not mutate the array in place.

Reply:
`slice` does not mutate, `splice`(which was used) does mutate, but in this case it's fine because the array was cloned before

Reply:
Oh yes, `splice` does... Yeah that's why I cloned it.